### ISSUE TITLE
Stuck on "FinishSpeech" after first command

### ISSUE THREAD

**Issue by Torsten85** [plamoni/SiriProxy#143]
Hey,
when I start the server, the first command is processed just fine. The second command stops at "[Info - iPhone] Received Object: FinishSpeech". So I have to cancel (on iPhone) and retry.

Is this an issue of the SiriProxy or is is a temporary problem of the apple server?

Thanks in advance
Torsten

**Comment by rckcrlr** [comment]
I'm having a similar issue with my SiriProxy-AppController plugin.  It's fairly consistent after the first command, but does happen after 4 or 5 commands as well.  Could be my code, but I'm finishing with "request_completed" in every listen loop.

**Comment by Torsten85** [comment]
in addition to that, I often run into the issue that "say" and "request_completed" have no effect at all. I see "say" and "request_completed" in the cli output, but my Siri is silent and keeps spinning on me.

The "first" SiriProxy (old code) wasn't that elegant to code but got the job done... hopefully someone will look into that...

**Comment by mrmosconi** [comment]
Just a "me too" post. I'd say about 30% of the time, it just stops after FinishSpeech, and the purple mic spins. If I manually stop the spinning, and resubmit, it usually works.

**Comment by muhkuh0815** [comment]
same problem here. and quite often, lets say ever 3-5 tries, sometimes 3 times in a row.

when i start siriproxy server -l  i see that they are just doing an endless  "ping pong"-game forwarding 9 bytes

have this problem with the old siriproxy version on mac, the new one on my Mac, and the new one on an VM Ubuntu.

**Comment by dcinzona** [comment]
Is this with the "text siri proxy" command or a custom plugin?

**Comment by Torsten85** [comment]
I've tested this with a fresh installation of the server and it also occurred with the test plugin.

By the way: I'm in germany so my Siri is set to german...

**Comment by Torsten85** [comment]
That's how it looks like in debug cli output:

```
[Debug - iPhone] Forwarding 9 bytes of data to Guzzoni
========UNZIPPED DATA (from Guzzoni =========
02 00 00 00 e9 62 70 6c 69 73 74 30 30 d5 01 03 07 09 0b 02 04 08 0a 0c 55 63 6c 61 73 73 5f 10 10 52 65 71 75 65 73 74 43 6f     6d 70 6c 65 74 65 64 5a 70 72 6f 70 65 72 74 69 65 73 d1 05 06 59 63 61 6c 6c 62 61 63 6b 73 a0 55 61 63 65 49 64 5f 10 24 66 62 31 66 32 36 39 33 2d 33 66 66 30 2d 34 34 33 34 2d 39 34 30 37 2d 66 63 36 35 65 64 62 30 65 62 63 31 55 72 65 66 49 64 5f 10 24 38 41 35 42 46 37 37 35 2d 32 31 44 33 2d 34 35 45 43 2d 39 32 43 42 2d 38 39 44 46 39 43 44 34 45 41 30 45 55 67 72 6f 75     70 5f 10 14 63 6f 6d 2e 61 70 70 6c 65 2e 61 63 65 2e 73 79 73 74 65 6d 08 13 19 2c 37 3a 44 45 4b 72 78 9f a5 00 00 00 00 00 00 01 01 00 00 00 00 00 00 00 0d 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 bc 04 00 00 00 10
==================================================
[Info - Dropping Object from Guzzoni] RequestCompleted
{"class"=>"RequestCompleted",
 "properties"=>{"callbacks"=>[]},
 "aceId"=>"fb1f2693-3ff0-4434-9407-fc65edb1ebc1",
 "refId"=>"8A5BF775-21D3-45EC-92CB-84DF9CD4CA0E",
 "group"=>"com.apple.ace.system"}
[Pong - Guzzoni] (16)
[Debug - Guzzoni] Forwarding 9 bytes of data to iPhone
========UNZIPPED DATA (from iPhone =========
03 00 00 00 11
==================================================
[Ping - iPhone] (17)
[Debug - iPhone] Forwarding 9 bytes of data to Guzzoni
========UNZIPPED DATA (from Guzzoni =========
04 00 00 00 11
==================================================
[Pong - Guzzoni] (17)
[Debug - Guzzoni] Forwarding 9 bytes of data to iPhone
========UNZIPPED DATA (from iPhone =========
03 00 00 00 12
```

Ping / Pong repeats indefinitly...

That's really the most annoying bug. SiriProxy is such a great app... but it's unusable atm.

**Comment by dcinzona** [comment]
That's interesting because it doesnt do that to me.  I wonder if it's because you're not in the US.  Was Germany one of the initially supported languages or locations or was this "Beta" US only?

I'd be interested to see if it's a regional thing vs a code thing.  I mean, the proxy works for many many people, so it has to be something with your setup or your location.  The code seems to function fine.

**Comment by muhkuh0815** [comment]
German lanuage was initially supportet, but i have seen this problem in some siriproxy plugin video (from england or australian, dont remember)  too. so its maybe an outside US problem.

edit: and yes we dont have WolframAlpha, Yelp restaurant search, nor anything including traffic/maps. but the test map example works

**Comment by Torsten85** [comment]
German was supported from day one (without WolframAlpha). It's working like it should without the SiriProxy... so the Servers for German-Siri are operational. 

Is Apple server interaction necessary when I use the "say" Function? Currently my commands are understood by the Proxy but neither the say function nor completed_request are send to my iPhone... maybe it's the same bug...

**Comment by mrmosconi** [comment]
I'm in the U.S.  Same problem. Both on Mac and Ubuntu.

**Comment by Torsten85** [comment]
I tested this behavior a little bit more and it seems like this only happens after a plugin command.
When a native command works, the following commands works as well.
But after a plugin command, there is a chance that you get stuck on "FinishSpeech". 

I'm unable to say "Test Siri Proxy" 10 times without getting stuck. But I can ask for the current time as often as I like.

I don't know much about the under-the-hood server communication, but it seems like the connection is not properly terminated (and yes, I use request_completed).

Could someone else please test this behavior?
I really wan't this to be solved :-)

**Comment by dcinzona** [comment]
I'm not seeing this problem.  I've had siriproxy running for 3 days (never restarted) and I have a custom plugin which still works fine (as does the test plugin).  Memory spiked to 18mb (not much) and then dropped again to 15-17mb but I doubt that would be related at all to your problem.  

I'm using 
Ruby 1.9.3p0, 
openSSL 1.0.0e or whatever, 
RubyGems 1.8.12, 
dnsmasq 2.52, 
siriproxy (build from 3 days ago).
Running on BusyBox linux (Synology NAS)

**Comment by gotnull** [comment]
I played with SiriProxy for hours last night and every single time it would freeze at FinishSpeech meaning I would have to say something just to kick her out of that state.
1. Test Siri proxy
   (she responds)
2. Test Siri proxy
   (FinishSpeech)
   (no response)

It's only until I tap the button and say something again that she starts to respond again.

I should mention I'm running Ubuntu 11.10 on my laptop and I'm located in Melbourne, Australia. But I recall this happening on my Mac running Snow Leopard and Lion (as I installed this thinking that it would fix the issue and it didn't).

Very annoying.

**Comment by Symo** [comment]
Just a ME TOO addition to this.  I'm running SiriProxy on latest Ubuntu in VM Fusion under OS X Lion.

Hopefully someone will have some insight on this soon.

Thanks.

**Comment by ScottJD** [comment]
This also is happening with mine. dnsmasq running on NAS and proxy running on Mac.

**Comment by martinmansson** [comment]
The same thing happening for me. It seems like about every second plugin command will freeze at 

[Info - iPhone] Received Object: FinishSpeech

Tapping the button and say something again and she will respond then the command after that will most likely make it freeze again.

Really annoying. A fix for this would be awesome :D

Thanks for great work!

**Comment by gotnull** [comment]
Yeah, same happens with me. Someone really needs to be assigned to fix this issue.

**Comment by keyosk** [comment]
Confirming this issue as well, english, in the us, with a centos based operating system.

**Comment by willcosgrove** [comment]
Yeah I'm getting this a lot also.  If you need any log files let me know.

**Comment by ninja98** [comment]
Yes, confirming the issue. Same here, running Mac..

**Comment by olehs** [comment]
Confirm

**Comment by Symo** [comment]
Moved from Ubuntu virtual machine to OS X Lion.  Still have same issue.

**Comment by ninja98** [comment]
I just removed some 'request_completed' lines and it seems to have fixed my problems with this.

**Comment by martinmansson** [comment]
ninja98 - Where did you remove them? From your own methods?

**Comment by ninja98** [comment]
Yes just commented the finish statements. Dont know if this causes
some other problem but I dont see any spinning anymore and it seems to work ok for me.

**Comment by gotnull** [comment]
@ninja98 So you're saying you're removing 'request_completed' calls after your 'say' commands?

**Comment by ninja98** [comment]
yes somehow it seems to work for me. you need at least one say though otherwise it will keep spinning.

**Comment by gotnull** [comment]
@ninja98 Unfortunately doesn't seem to work for me at all. I'm only ever using one `request_completed` call any way and I still get spinning/freezes at `FinishSpeech`.

**Comment by ghost** [comment]
I'm adding my name to this list as well.  I've been thinking it was my code the whole time.

**Comment by ghost** [comment]
Ok guys, it seems to be like this:

I've tested something on my Proxy.

It seems like if you trying to ask Siri something that can be interpreted with "search for "name" " or something Siri is able to work with, that your Proxy @ home (if this action you asked for or you commanded is processed by a plugin on the Proxy), the Guzzoni-Server is still trying to get an answerer from your iPhone.
That's when you get those endless Ping-Pong-Sessions.

For Example:
I edited the example-plugin.
Listen for was set to "Hello I am Christopher"
Package was successfully transmitted to Guzzoni
Package was successfully received from Guzzoni
Plugin starts
Guzzoni is trying to find contact "Christopher" (PING PONG PING PONG)
.....

But your Request did finish (but for SiriProxy only...)

Am i right? Or did i miss something... pls check that.

**Comment by gotnull** [comment]
@DJstone Looks like you're on the right track.

**Comment by BenWoodford** [comment]
Quick "Me too" comment. Seems to be rather random as to when it happens. Running on an Ubuntu 32-Bit VM on an ESXi host, server itself is gigabit so it's not going to be a lag problem and it happens on both WiFi and 3G.

On a 4S by the way.

**Comment by gotnull** [comment]
This is still happening. Just to let you all know. ;-)

**Comment by TobyTech** [comment]
"Me too."

My iPhone works flawlessly, but my iPad hangs at Finish Speech.

**Comment by hilko** [comment]
Same issue here, on an iPad 2 from Germany, but using a UK-based VPS. Happens consistently.

**Comment by TobyTech** [comment]
iPhone over 3G to SiriProxy ---->  WORKS
iPhone over local WiFi to SiriProxy ----> WORKS

iPad over 3G/OtherWiFi to SiriProxy ----> FINISHSPEECH!
IPad over local WiFi to SiriProxy  ----> FINISHSPEECH!

Conclusion:

IP wasn't banned, UDID Banning??????!?!?!?!?!   <--- Going to do more research by spoofing UDID.

Error with non-4S devices with westbaer's fork?  <----- NOT TRUE, iPad worked before!

**Comment by muhkuh0815** [comment]
if you say "test siri proxy" it wont hang, but with "test map" it will very often, on 4, 4s and ipad2

**Comment by sturma** [comment]
I'm also experiencing this. If a command needs Siriproxy to connect the internet or any IP element, the next command hangs at "FinishSpeech".
Ubuntu 32-bit or 64-bit, 11.10 or 12.04.1, VM on Virtualbox or directly on PC.
Has anyone managed to solve it or use a workaround?

**Comment by wmacevoy** [comment]
Ditto: Ubuntu 12.04 on VM Fusion -- "Mari turn on the light" (works) followed by "Mari turn off the light" (hangs).  iphone 4s.  I only get one command in before a hang.

**Comment by sturma** [comment]
@dcinzona Are you still using the same configuration or have you updated ruby, rubygems etc. to latest releases?
thanks

**Comment by Alsaegh** [comment]
wow ,, i thought it was only me and i was thinking its my code !! .. am having the same issue but i am adding my custom commands to the original example file that comes in plugin folder with siriproxy ,, not only this 

its even when it stuck  i try to click and type my voice command to make it match exact phrase, it didn't recognize my custom command .. it only recognize it from first voice try, and when try to click and type, it failed to match with my custom commands

**Comment by sturma** [comment]
I noticed this problem occurs even when only the example-plugin is enabled in config.yml.
i.e. "test siriproxy" > getting reply > "siriproxy test question" > getting second question > "yes" > getting reply > "siriproxy test question" > hanging...

**Comment by Alsaegh** [comment]
No solution for this problem ? :(

**Comment by wmacevoy** [comment]
This is a terrible hack, but I have a working solution.  I throw an exception 1 second after completing a response, then I have a wrapper script that watches for that exception exit and immediately restarts the service.  Siriproxy reboots fast (<1 second), and I don't talk that fast, so it works.  But, like I said, it's a terrible hack.

Here's the part of the plugin (controlling a snoopy lawn ornament) that throws the exception:

``` ruby
  listen_for /check.*(snoopy)/i do |noun|
    is_on = ctrl(noun,"on?")
    say noun + " is " + (is_on ? "on" : "off") + "."
    request_completed
    restart
  end

  def restart
    print "restarting server..."
    Thread.abort_on_exception = true
    Thread.new do
      sleep 1
      raise "restart"
    end
  end
```

And here's the bash script that runs the service (on a mac, with small modifications on ubuntu)

``` bash
#!/bin/bash                                                                     

cd /usr/local/SiriProxy || exit 1
. /usr/local/rvm/scripts/rvm


#                                                                               
# look for "restart" in last few lines of siriproxy output                      
#                                                                               
export RESTART

restarter() {
  count=0
  while read -r line
  do
    if [[ "$line" =~ restart ]]
    then
      count=0
    else
      count=$((count+1))
    fi
    echo "$line"
  done
  echo $((count<5)) > "$RESTART"
}

on_exit() {
  /bin/rm -rf "$RESTART"
}

RESTART=$(mktemp -t siriproxy)
trap on_exit EXIT

echo 1 > $RESTART
while [ $(cat $RESTART) -ne 0 ]
do
  echo 0 > $RESTART
  siriproxy server 2>&1 | restarter
done
```

So long as you are permanently storing state and the reboots are fast, this works.

**Comment by sturma** [comment]
@wmacevoy: Will this script allow SiriProxy to retry the last command or simply prevents the process from stalling?

**Comment by Raggyx** [comment]
Hi
I do not know what caused the error, but I have found a way to fix this. works fine for me.
Edit the File connection.rb
go to the function prep_received_object(object) and insert the following lines directly below the function name
<code>
def prep_received_object(object)
</code>
Add this:
<code>
    if object["class"] == "FinishSpeech" or object["class"] == "SpeechRecognized"
        @block_rest_of_session = false
    end

</code>

let me know if that worked for you

**Comment by sturma** [comment]
Hi,

It didn’t work for me… do I need to make a “rake install” afterwards? Only did “siriproxy bundle” and “bundle install” and restarted the process.

The file is located under ~/siriproxy/lib/siriproxy, right?

**Comment by Raggyx** [comment]
the file is located under ~/gems/siriproxy-0.11.3/lib/siriproxy
you only need siriproxy bundle. All updates and reinstall will overwrite the changes.

**Comment by sturma** [comment]
Ok, my path was a bit different, but it did the trick! It worked! Continuous commands without getting stuck at all!

/home/alex/.rvm/gems/ruby-1.9.3-p286@SiriProxy/gems/siriproxy-0.11.3/lib/siriproxy

Thank you : )

**Comment by naek** [comment]
@Raggyx: Nice find! :+1:  I too, found a solution simliar to yours.  Not sure why it is stalling though, it seem that the speech does not get processed into plain text correctly and just hangs there.  This happens with iPhone 5 & 4S.

**Comment by Alsaegh** [comment]
@Raggyx: Your Solution worked for me :+1: 
Thanks a lot =)

**Comment by elvisimprsntr** [comment]
fixed in 0.5.3
